codegen-service/llm-generation-core/src/main/java/org/rj/modelgen/llm/prompt/TemplatedPromptGenerator.java

```diff
@@ -0,0 +1,54 @@
+package org.rj.modelgen.llm.prompt;
+
+import java.io.StringReader;
+import java.io.StringWriter;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import freemarker.template.Configuration;
+import freemarker.template.Template;
+
+public class TemplatedPromptGenerator<TImpl extends TemplatedPromptGenerator<?, TSelector>, TSelector> extends PromptGenerator<TImpl, TSelector> {
+    private final Map<TSelector, String> prompts;
+
+    public TemplatedPromptGenerator() {
+        this(Map.of());
+    }
+
+    public TemplatedPromptGenerator(Map<TSelector, String> prompts) {
+        this.prompts = new HashMap<>(prompts);
+    }
+
+    @Override
+    public TImpl withAvailablePrompt(TSelector selector, String prompt) {
+        return super.withAvailablePrompt(selector, prompt);
+    }
+
+    @Override
+    public Optional<String> getPrompt(TSelector selector, List<PromptSubstitution> substitutions) {
+        return Optional.ofNullable(selector)
+                .map(prompts::get)
+                .map(raw -> {
+                    Configuration configuration = new Configuration(Configuration.VERSION_2_3_23);
+                    configuration.setClassForTemplateLoading(configuration.getClass(), "/");
+                    try {
+                        Template template = new Template("template", new StringReader(raw), configuration);
+                        StringWriter output = new StringWriter();
+                        template.process(createDataModel(parameters), output);
+                        return output.toString();
+                    } catch (Exception ex) {
+                        return null;
```

Comment:
If we fail at this point then we've located the relevant prompt template and tried to generate it, but something blew up while we were doing it.  It's probably better to fail loudly in this case.  I'd recommend throwing a runtime exception like `LlmGenerationModelException` and include `ex` as the cause, which should cause the overall generation process to fail.

Does the FTL generator declares that it may throw certain specific exceptions?   If so then I'd consider handling those in one `catch` (with a more specific error message) and then handle the `Exception` as a fallback in a second catch.  They can both throw the same type of runtime exception though

Reply:
Addressed

Reply:
It would be good if you can include the original exception as the 'cause' for traceability too, e.g. `throw new XException(ex.getMessage(), ex)`

Reply:
Addressed